### PR TITLE
Move ssc-node to the @socketsupply/socket

### PR DESCRIPTION
**Pull request by chicoxyzzy**
## Allow CommonJS

Experiment with npm workspaces and module exports. This allows to symlink the `@socketsupply/socket` from the test app. I moved `ssc-node` to the `node` directory of `@socketsupply/socket`. I use `ssc-node` and not https://github.com/socketsupply/socket/pull/133 to be sure all tests are passing. This patch in `@socketsupply/socket`'s `package.json` allows us to enable usage of both CommonJS and ESM

```json
  "exports": {
    "import": "./node/index.js",
    "require": "./node/index.cjs"
  },
```

CommonJS **is not possible** in the https://github.com/socketsupply/socket/pull/133 draft.

This approach also allows us to split the frontend IPC and backend IPC and reuse the same modules. If necessary they can share the base class, but we will need to move the `/api` files to the `/npm/packages/@socketsupply/socket` (what in fact looks like a proper thing to do to me anyway! cc @jwerle)

# Disadvantages of moving Node.js code to the `api`

1. One **big** disadvantage is that `@socketsupply/socket`+ it's platform-specific dependancy are a very heavyweight packages and they will be added to **each** Socket Runtime project even if Socket Runtime already installed globally. For `darwin-x64` it's almost 18 MB and for `darwin-arm64` it's almost 28 MB!

1. Another disadvantage is that imports are different for frontend usage and backend usage.
```js
// frontend
import { ipc } from 'socket:ipc'
```
```js
// backend
import { ipc } from '@socketsupply/ipc'
```

That looks like **two different dependancies**. Import maps are not supported in Node.js yet, so we can't make it `socket:ipc` there ([Deno has](https://deno.land/manual@v1.30.0/basics/import_maps) import maps support). 

# Summary

Given these two disadvantages, I'd prefer to have a **separate tiny module** named `@socketsupply/socket-ipc-node` (ex-[`@socketsupply/ssc-node`](https://github.com/socketsupply/ssc-node)) for Node.js. This module also should mirror the `socket:ipc` API and have both ESM and CommonJS support.

### CODE REVIEW

File: npm/packages/@socketsupply/socket/node/index.js [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    process.stderr.write('RAW MESSAGE: ' + s.slice(0, 512) + '...\n')
+  }
+
+  return new Promise(resolve =>
+    process.stdout.write(s + '\n', resolve)
+  )
+}
+
+console.log = (...args) => {
+  const s = args.map(v => format(v)).join(' ')
```
Comment (jwerle) [inline npm/packages/@socketsupply/socket/node/index.js:29]: Format string in first argument is dropped.

Try

```js
util.format(...args)
```

File: npm/packages/@socketsupply/socket/node/index.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+  const enc = encodeURIComponent(s)
+  // fs.appendFileSync('tmp.log', s + '\n')
+  write(`ipc://stdout?value=${enc}`)
+}
+console.error = console.log
```
Comment (jwerle) [inline npm/packages/@socketsupply/socket/node/index.js:34]: There is a `ipc://stderr` route

File: npm/packages/@socketsupply/socket/node/index.js [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+// Internal IPC API
+//
+const ipc = { nextSeq: 0 }
+
+ipc.resolve = async (seq, state, value) => {
+  const method = !Number(state) ? 'resolve' : 'reject'
+  if (!ipc[seq] || !ipc[seq][method]) return
+
+  /**
+   * TODO Who handles this error ?
```
Comment (jwerle) [inline npm/packages/@socketsupply/socket/node/index.js:66]: Probably `ipc` if it was an event emitter!

File: npm/packages/@socketsupply/socket/node/index.js [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+  }
+
+  await write(`ipc://${cmd}?${value}`)
+  return promise
+}
+
+ipc.send = async o => {
+  try {
+    // TODO: use structuredClone instead once we are on node 17+
+    o = JSON.parse(JSON.stringify(o))
```
Comment (jwerle) [inline npm/packages/@socketsupply/socket/node/index.js:111]: You'd be surprised to know that `structuredClone` is much slower [1]

1. https://www.measurethat.net/Benchmarks/Show/18541/0/jsonstringify-vs-structuredclone#latest_results_block

**Comment by chicoxyzzy** [comment]
@jwerle code in the `node` repo is just a copy of `ssc-node`. I don't want to move it to the `socket` repo. Here I did it in order to try how npm workspaces work. I don't want to merge this branch. Although, I think that we should move `/api` files to the `/npm/packages/@socketsupply/socket` repo. Please read the issue description.

**Comment by jwerle** [comment]
> @jwerle code in the `node` repo is just a copy of `ssc-node`. I don't want to move it to the `socket` repo. Here I did it in order to try how npm workspaces work. I don't want to merge this branch. Although, I think that we should move `/api` files to the `/npm/packages/@socketsupply/socket` directory. Please read the issue description.

yes, sorry. I clearly went into auto review mode without finishing

I don't think we should move the `api/` source files into the `npm/` directory. The NPM module should be considered a superset of the runtime js api.

**Comment by chicoxyzzy** [comment]
Ok, I think we have consensus: we want a new module that will mirror the API from #133. We can put this module in the monorepo to make it easier to introduce changes along with `socket:ipc` and test app updates 

I'm going to close this PR

**Pull request closed without merging**